### DISCUSSION TITLE
When running multiple instances connected to the same database, is it possible for data races to occur?

### DISCUSSION THREAD

**Discussion by tmscer** [directus/directus#11303] [Archive]
When running multiple instances connected to the same database (in my case Postgres), is it possible for data races to occur? Does using cache (e.g. Redis) have an effect? Are transactions always used for operations that require multiple SQL statements?

I of course know that Postgres checks data against schema but there are tables whose fields are JSON which is not checked by Postgres.

Thanks for answers or any pointers, I was unable to find out on my own and.

**Comment by rijkvanzanten** [comment]
> is it possible for data races to occur

It depends on exactly what scenario you have in mind when you mention _data races_, but technically speaking, yes multiple servers could inject into the same table at the same time. That being said, it's always just a single Directus instance that does the currently requested read/write operation, so you won't have any risk of Directus inserting the same item twice on a single request. Any operation that requires multiple queries, like nested relational insertion, is done in transactions. 

> Does using cache (e.g. Redis) have an effect?

Yes, if you run Directus horizontally scaled, you should rely on a shared cache engine to ensure you don't have mismatches between caches between the instances of Directus. A single Redis instance that is connected to every Directus instance is the recommended approach.
